module/zfs/mmp.c

```diff
@@ -125,27 +130,79 @@ mmp_thread_stop(dsl_pool_t *dp)
 	mmp->mmp_thread_exiting = 0;
 }
 
+vdev_t *
+vdev_random_leaf(spa_t *spa)
+{
+	vdev_t *vd, *child;
+
+	ASSERT(spa);
+	ASSERT(spa_config_held(spa, SCL_STATE, RW_READER) == SCL_STATE);
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+		if (!vdev_is_dead(child))
+			vd = child;
+	}
+
+	return (vd);
+}
+
 static void
 mmp_thread(dsl_pool_t *dp)
 {
 	mmp_thread_state_t *mmp = &dp->dp_mmp;
-	clock_t start;
-	/*
 	spa_t *spa = dp->dp_spa;
-	*/
+	abd_t *abd;
+	char *buf;
+	int bufsize = (1ULL << MAX_UBERBLOCK_SHIFT);
+
+	abd = abd_alloc(bufsize, B_TRUE);
+	buf = kmem_alloc(bufsize, KM_SLEEP);
+	bzero(buf, bufsize);
 
 	mmp_thread_enter(mmp);
 
 	dprintf("entered: pool %p\n", dp);
 
 	mmp->mmp_last_update = gethrtime();
 	for (;;) {
+		int flags = ZIO_FLAG_CONFIG_WRITER | ZIO_FLAG_CANFAIL;
+		clock_t start;
+		int l, n, rc;
+		vdev_t *vd;
+		zio_t *zio;
+
 		if (mmp->mmp_thread_exiting)
-			mmp_thread_exit(mmp, &mmp->mmp_thread);
+			goto cleanup_and_exit;
 
 		start = gethrtime();
 
-		/* Do actual mmp reads/writes here */
+		/* Copy the latest uberblock and update the time */
+		(void) memcpy(buf, &spa->spa_ubsync, sizeof (uberblock_t));
+		((uberblock_t *)buf)->ub_timestamp = gethrestime_sec();
+		abd_copy_from_buf(abd, buf, bufsize);
+
+		/*
+		 * Choose a random vdev and MMP block, and write over it in all
+		 * labels.  XXX need to retry if the zio fails
+		 */
+		spa_config_enter(spa, SCL_STATE, FTAG, RW_READER);
+		vd = vdev_random_leaf(spa);
+		if (vd) {
+			n = spa_get_random(MMP_BLOCKS_PER_LABEL);
```

Comment:
On linux, spa_get_random() calls an spl function random_get_pseudo_bytes() which implements its own PRNG instead of using the kernel-provided one.   Therefore this does not use up entropy gathered by the kernel.

I did alter spa_get_random() to detect range==1 and return 0 immediately.